### PR TITLE
#41, #36

### PR DESCRIPTION
**Pull request by ivmarkov**
As the subject says, this is fixing #41 and #36.

The changes should be relatively straightforward actually, with the non-trivial changes actually being part of the `espidf` module in `embuild`. 

See this PR for the `embuild`-related changes:
https://github.com/ivmarkov/embuild/pull/29

### CODE REVIEW

**Comment by N3xed** [comment]
Okay, I think this is close to finished.
Still needs some more testing though, and code reviews are appreciated.

I've refactored @ivmarkov 's code pretty heavily and also incorporated the changes I've made in `embuild` (though the author of the first commit is still, ivmarkov; hope you don't mind).

The biggest change is that `EspIdf` now uses a closure for the tools to be installed, this allows to decide dynamically which tools should be installed. This change together with the added `EspIdfVersion` that is parsed from the `esp-idf/tools/cmake/version.cmake` file now allows us easily to very easily support multiple esp-idf versions.

I've also added a big `match` statement for selecting the esp-idf detected from the environment, the repo specified with `IDF_PATH` or everything managed by the installer.

Basically:
- If activated esp-idf available and $ESP_IDF_TOOLS_INSTALL_DIR == "fromenv" or unset, use it
- If $IDF_PATH is valid, use it as user provided esp-idf repository (but installer manages tools)
- Otherwise the installer manages everything and respects `ESP_IDF_REPOSITORY` and `ESP_IDF_VERSION`.

There are also some error and warnings for edge-cases.
